godot-macros/src/docs.rs

```diff
@@ -0,0 +1,151 @@
+use crate::class::{ConstDefinition, FuncDefinition};
+use proc_macro2::TokenStream;
+use quote::ToTokens;
+use regex::Regex;
+use std::sync::OnceLock;
+use venial::*;
+
+/// Converts [markdown](https://en.wikipedia.org/wiki/Markdown) to [BBCode](https://en.wikipedia.org/wiki/BBCode).
+pub fn md2bb(md: &str) -> String {
+    macro_rules! r {
+        ($($expr:literal => $replace:literal)+) => {{
+            const REGEXES: OnceLock<[(Regex, &str); 6]> = OnceLock::new();
+            REGEXES.get_or_init(|| [
+                $((Regex::new($expr).unwrap(), $replace)),+
+            ])
+        }}
+    }
+    // this is not the most robust way
+    let mut out = md.to_string();
+    for (search, replace) in r![
+        "_([^_]+)_" => "[i]$1[/i]"
+        "\\*\\*([^\\*\\*]+)\\*\\*" => "[b]$1[/b]"
+        "\\*([^\\*]+)\\*" => "[i]$1[/i]"
+        "```([^`]+)```" => "[codeblock]$1[/codeblock]"
+        "`([^`]+)`" => "[code]$1[/code]"
+        "~~([^~]+)~~" => "[s]$1[/s]"
+    ] {
+        out = search.replace_all(&out, *replace).into_owned();
+    }
+    out
+}
+
+pub fn constant(cnst: Constant) -> String {
+    let name = cnst.name.to_string();
+    let value = cnst
+        .initializer
+        .map(|x| x.to_token_stream().to_string())
+        .unwrap_or("null".into());
+    let docs = docs(cnst.attributes);
+    format!(r#"<constant name="{name}" value="{value}">{docs}</constant>"#)
+}
+
+/// This will register the documentation.
+pub fn class(
+    functions: impl IntoIterator<Item = FuncDefinition>,
+    constants: impl IntoIterator<Item = ConstDefinition>,
+) -> TokenStream {
+    let methods = functions.into_iter().map(method).collect::<String>();
+
+    let constants = constants
+        .into_iter()
+        .map(|ConstDefinition { raw_constant: x }| x)
+        .map(constant)
+        .collect::<String>();
+
+    let init = format!(
+        r#"
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="CLASS⯂" inherits="BASE⯂"
+  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../class.xsd">
+  <brief_description>BRIEF⯂</brief_description>
+  <description>DESC⯂</description>
+  <methods>
+    {methods}
+  </methods>
+  <constants>
+    {constants}
+  </constants>
+</class>
+    "#
+    );
+
+    quote::quote! {
+        use ::godot::prelude::GodotClass;
+        let xml = #init
+            .replace("CLASS⯂", Self::class_name().as_str())
+            .replace("BASE⯂", Self::__BASE.unwrap_or("Object"))
+            .replace("BRIEF⯂", Self::__DOCS.lines().next().unwrap_or_default())
+            .replace("DESC⯂", Self::__DOCS);
+        ::godot::DOCS.lock().unwrap().push(xml);
+    }
```

Comment:
this seems like a pretty fragile solution

Reply:
i think this is one of the more okay parts of the code.
the solution to this would include a static inside the proc macro, which, well.
if you're talking about the last line though, i agree.

Reply:
i dont think we should use effectively magic strings that can be replaced, but rather construct the xml up while interpolating on the way